include/verilated_vcd_c.cpp

```diff
@@ -804,6 +874,266 @@ void VerilatedVcd::fullBusX(vluint32_t code, int bits) {
+    *oldp = newval;
+    char* wp = m_writep;
+    *wp++ = 'b';
+    newval <<= 32 - N;
+    int bits = N;
+    do {
+        *wp++ = '0' | static_cast<char>(newval >> 31);
+        newval <<= 1;
+    } while (--bits);
+    finishLine(oldp, wp);
+}
+// Note: No specialization for width 1, covered by 'fullBit'
+template void VerilatedVcd::fullBus<2>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<3>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<4>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<5>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<6>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<7>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<8>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<9>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<10>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<11>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<12>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<13>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<14>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<15>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<16>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<17>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<18>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<19>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<20>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<21>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<22>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<23>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<24>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<25>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<26>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<27>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<28>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<29>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<30>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<31>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedVcd::fullBus<32>(vluint32_t* oldp, vluint32_t newval);
+
+// N is the number of used bits in the value
+template <int N> void VerilatedVcd::fullQuad(vluint32_t* oldp, vluint64_t newval) {
+    *reinterpret_cast<vluint64_t*>(oldp) = newval;
```

Comment:
I wonder if the fullQuads are worth it.  Maybe try this:

    WData w[VL_WQ_WORDS_E]; VL_SET_WQ(rhsw, newval);

Then calling fullArray(old, w, 1). My thinking is it is only a few instructions different and a lot less code to compile/keep in cache.  (With the old scheme it was a lot more overhead.)

Reply:
So the `chgQuad` definitely is worth it beacuase of the single 64-bit compare vs two 32-bit ones. I changed `fullQuad` to do what you suggest and call `fullArray`, but that made it clearly 1% slower. SweRV is a 32-bit machine so it actually uses Quads very rarely (only 7 of the 32 variants actually show up in the profile). My worry is that on a wider design, this would hurt more.

I also was wondering about the code size impact of these templates. On a more point I think we should build everytihng (or at least the hand-written support files in `include/`) with `-ffunction-sections` and link with `-Wl,-gc-sections`, I don't think this would impact model build time noticably, but could reduce ITLB misses, though I don't expect much.

In this particular case, we can change the quad/array varriant to be tempalted based on the number of bits mod some smaller constant say 4, and then hand unroll the loops by that amount, handling the remainder bits based on teh template (this is basically what clang compiles those into, except the main loop would now be variable length). I will try this shorty.

Reply:
Keeping the quads will be worth it then.

Maybe compare to something like this:
```
    // clang-format off
    switch (size) {
    case 31: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 30: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 29: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 28: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 27: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 26: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 25: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 24: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 23: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 22: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 21: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 20: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 19: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 18: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 17: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 16: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 15: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 14: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 13: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 12: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 11: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 10: *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 9:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 8:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 7:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 6:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 5:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 4:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 3:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 2:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 1:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    case 0:  *wp++ = '0' | static_cast<char>(newval >> 31); newval<<1; //FALLTHRU
    }
    // clang-format on

```

Reply:
(Basically one hard to predict branch, then that's it.)

Reply:
(Meant <<= 1 above here, will edit.)

Reply:
I would be reluctant to use -ffunction-sections by default given how GCC warns about it slowing linking.

Reply:
That branch table seems to work quite well, so I reverted `fullQuad` and `fullArray` to no use templates.